File: juego_ahorcado.kt
import kotlin.random.Random

// # Primer punto: implementacion de clases manual 

class EmptyStackException(message: String) : Exception(message)

class StackManual<T> {
    private val items = mutableListOf<T>()

    fun push(value: T) = items.add(value)

    fun pop(): T {
        if (isEmpty()) throw EmptyStackException("La pila está vacía (no hay jugadas para deshacer).")
        return items.removeAt(items.size - 1)
    }

    fun peek(): T {
        if (isEmpty()) throw EmptyStackException("La pila está vacía.")
        return items.last()
    }

    fun isEmpty() = items.isEmpty()

    fun size() = items.size

    fun clear() = items.clear()
}


class EmptyQueueException(message: String) : Exception(message)

class QueueManual<T> {
    private val items = mutableListOf<T>()

    fun enqueue(value: T) = items.add(value)

    fun dequeue(): T {
        if (isEmpty()) throw EmptyQueueException("La cola está vacía.")
        return items.removeAt(0)
    }

    fun front(): T {
        if (isEmpty()) throw EmptyQueueException("La cola está vacía.")
        return items[0]
    }

    fun isEmpty() = items.isEmpty()

    fun size() = items.size

    fun clear() = items.clear()
}


// Imporporacion de colas y pilas al juego ahorcado


fun obtenerPalabraAleatoria(): String {
    val palabras = listOf("materia", "comida", "agua", "gato", "felicidad", "saltar", "futbol")
    return palabras[Random.nextInt(palabras.size)]
}

fun mostrarTablero(palabraSecreta: String, letrasAdivinadas: List<Char>) {
    val tablero = buildString {
        for (letra in palabraSecreta) append(if (letra in letrasAdivinadas) letra else '_')
    }
    println(tablero)
}

fun jugarAhorcado() {
    val palabraSecreta = obtenerPalabraAleatoria()
    val letrasAdivinadas = mutableListOf<Char>()
    val pilaUndo = StackManual<Char>()        // Pila para deshacer jugadas
    val colaJugadores = QueueManual<String>() // Cola para turnos
    colaJugadores.enqueue("Jugador 1")
    colaJugadores.enqueue("Jugador 2")

    var intentosRestantes = 10

    println("\n=== JUEGO DEL AHORCADO ===")
    println("Comando adicional: escribe UNDO para deshacer la última letra ingresada.")

    while (intentosRestantes > 0) {
        val jugadorActual = colaJugadores.dequeue()
        colaJugadores.enqueue(jugadorActual)

        println("\nTurno de: $jugadorActual")
        mostrarTablero(palabraSecreta, letrasAdivinadas)

        print("Pila pues! introduce una letra (o UNDO): ")
        val input = readLine()?.lowercase() ?: ""
        if (input.isEmpty()) continue

        if (input == "undo") {
            try {
                val ultima = pilaUndo.pop()
                letrasAdivinadas.remove(ultima)
                println("Se deshizo la letra: $ultima")
            } catch (e: Exception) {
                println(e.message)
            }
            continue
        }

        val letra = input[0]

        if (letra in letrasAdivinadas) {
            println("Ya habías ingresado esta letra cachon!.")
            continue
        }

        pilaUndo.push(letra)

        if (palabraSecreta.contains(letra)) {
            letrasAdivinadas.add(letra)
            if (palabraSecreta.toSet().all { it in letrasAdivinadas }) {
                println("¡Tu si sirves Ganaste! La palabra era: $palabraSecreta")
                return
            }
        } else {
            intentosRestantes--
            println("Esa no era cachon. Te quedan $intentosRestantes intentos.")
        }
    }

    println("No te dio el coco. La palabra era: $palabraSecreta")
}


// Segunta parte: version con codigos nativos

fun pilaNativaDemo() {
    val pila = ArrayDeque<Char>()
    pila.addLast('a')
    pila.addLast('b')
    pila.addLast('c')
    println("\nPila nativa: $pila")
    println("Pop (removeLast): ${pila.removeLast()}")
    println("Después: $pila")
}

fun colaNativaDemo() {
    val cola = ArrayDeque<String>()
    cola.addLast("Jugador 1")
    cola.addLast("Jugador 2")
    println("\nCola nativa: $cola")
    println("Dequeue (removeFirst): ${cola.removeFirst()}")
    println("Después: $cola")
}



// ALGORITMOS (SIN CAMBIOS)


fun busquedaLineal(lista: List<Int>, clave: Int): Int {
    for (i in lista.indices) if (lista[i] == clave) return i
    return -1
}

fun busquedaBinaria(lista: List<Int>, clave: Int): Int {
    var inicio = 0
    var fin = lista.size - 1
    while (inicio <= fin) {
        val medio = (inicio + fin) / 2
        when {
            lista[medio] == clave -> return medio
            lista[medio] < clave -> inicio = medio + 1
            else -> fin = medio - 1
        }
    }
    return -1
}

fun burbuja(lista: MutableList<Int>): List<Int> {
    val n = lista.size
    for (i in 0 until n - 1) {
        for (j in 0 until n - i - 1) {
            if (lista[j] > lista[j + 1]) {
                val temp = lista[j]
                lista[j] = lista[j + 1]
                lista[j + 1] = temp
            }
        }
    }
    return lista
}

fun insercion(lista: MutableList<Int>): List<Int> {
    for (i in 1 until lista.size) {
        val clave = lista[i]
        var j = i - 1
        while (j >= 0 && lista[j] > clave) {
            lista[j + 1] = lista[j]
            j--
        }
        lista[j + 1] = clave
    }
    return lista
}



// Ejecutar el codigo completo 



fun main() {
    jugarAhorcado()

    println("\n=== VERSIONES NATIVAS ===")
    pilaNativaDemo()
    colaNativaDemo()
}

File: juego_ahorcado.py
import random
from collections import deque

# Primer punto: IMPLEMENTACIÓN MANUAL

class EmptyStackException(Exception):
    pass

class StackManual:
    def __init__(self):
        self.items = []

    def push(self, value):
        self.items.append(value)

    def pop(self):
        if self.isEmpty():
            raise EmptyStackException("La pila está vacía, no se puede deshacer.")
        return self.items.pop()

    def peek(self):
        if self.isEmpty():
            raise EmptyStackException("La pila está vacía.")
        return self.items[-1]

    def isEmpty(self):
        return len(self.items) == 0

    def size(self):
        return len(self.items)

    def clear(self):
        self.items = []


class EmptyQueueException(Exception):
    pass

class QueueManual:
    def __init__(self):
        self.items = []

    def enqueue(self, value):
        self.items.append(value)

    def dequeue(self):
        if self.isEmpty():
            raise EmptyQueueException("La cola está vacía.")
        return self.items.pop(0)

    def front(self):
        if self.isEmpty():
            raise EmptyQueueException("La cola está vacía.")
        return self.items[0]

    def isEmpty(self):
        return len(self.items) == 0

    def size(self):
        return len(self.items)

    def clear(self):
        self.items = []



# Imporporacion de colas y pilas al juego ahorcado


def obtener_palabra_aleatoria():
    palabras = ["materia", "comida", "agua", "gato", "felicidad", "saltar", "futbol"]
    return random.choice(palabras)

def mostrar_tablero(palabra_secreta, letras):
    tablero = ""
    for letra in palabra_secreta:
        tablero += letra if letra in letras else "_"
    print(tablero)

def jugar_ahorcado():
    palabra_secreta = obtener_palabra_aleatoria()
    letras_adivinadas = []
    pila_undo = StackManual()  # Guardar letras ingresadas
    cola_jugadores = QueueManual()

    # Registrar jugadores
    cola_jugadores.enqueue("Jugador 1")
    cola_jugadores.enqueue("Jugador 2")

    intentos_restantes = 10
    print("\nJUEGO DEL AHORCADO")

    while intentos_restantes > 0:
        jugador_actual = cola_jugadores.dequeue()
        cola_jugadores.enqueue(jugador_actual)

        print(f"\nTurno de: {jugador_actual}")
        mostrar_tablero(palabra_secreta, letras_adivinadas)
        letra = input("Empieza cachon! Escribe una letra, o escribe UNDO para deshacer última letra: ").lower()

        if letra == "undo":
            try:
                ultima = pila_undo.pop()
                letras_adivinadas.remove(ultima)
                print(f"Se deshizo la letra: {ultima}")
            except:
                print("No hay jugadas por deshacer.")
            continue

        if letra in letras_adivinadas:
            print("Tu si eres cachon, ya habías ingresado esa letra.")
            continue

        pila_undo.push(letra)

        if letra in palabra_secreta:
            letras_adivinadas.append(letra)
            if set(palabra_secreta).issubset(set(letras_adivinadas)):
                print(f"¡Tu si sirbes ganaste! La palabra era: {palabra_secreta}")
                return
        else:
            intentos_restantes -= 1
            print(f"Esa no era cachon. Intentos restantes: {intentos_restantes}")

    print(f"No te dio el coco. La palabra era: {palabra_secreta}")



# Segunta parte: version con codigos nativos

def pila_nativa():
    pila = []  # list como pila
    pila.append("a")
    pila.append("b")
    pila.append("c")
    print("\nPila nativa:", pila)
    print("Pop:", pila.pop())
    print("Después:", pila)

def cola_nativa():
    cola = deque()  # deque como cola
    cola.append("Jugador 1")
    cola.append("Jugador 2")
    print("\nCola nativa:", cola)
    print("Dequeue:", cola.popleft())
    print("Después:", cola)


# Ejecutar el codigo completo 
if __name__ == "__main__":
    jugar_ahorcado()
    pila_nativa()
    cola_nativa()
